### PR TITLE
feat: implement profile management and restrictions

### LINKED ISSUE #880: Implement profile management in SPAR backend for Postgres
Describe the task:

We need to start planning out and implementing profile management in SPAR. PO will provide rules for Orchard_manager and TSC_Admin profiles.

These rules will need to be implemented on the back-end for Post Gres

Acceptance Criteria:
- [ ] authorization on back end 
- [ ]  roles only see allowed APIs and Endpoints
- [ ] tested and merged
- [ ] relevant documentation updated

### PR DESCRIPTION
**Pull request by RMCampos**
# Description
Closes #880 

#### New
- Interceptor class to verify required and existing access roles.

#### Changed
- All endpoint classes adding required roles;
- All endpoint test classes with proper required roles;

#### Removed
- None;

### How was this tested?
- [ ] 🧠 Not needed
- [x] 👀 Eyeball
- [x] 🤖 Added tests

###  What gif/image best describes this PR or how it makes you feel?

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://nr-spar-26-frontend.apps.silver.devops.gov.bc.ca/)
- [Backend](https://nr-spar-1076-backend.apps.silver.devops.gov.bc.ca/actuator/health)
- [Oracle-API](https://nr-spar-1076-oracle-api.apps.silver.devops.gov.bc.ca/actuator/health)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-spar/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge.yml)

### CODE REVIEW

**Review by mgaseta — APPROVED** [review]
If I have the right profile to say it: great job! :rocket: :1st_place_medal: 
LGTM!

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/FavouriteActivityEndpoint.java [hunk 1]
```diff
@@ -80,6 +83,18 @@ public class FavouriteActivityEndpoint {
...
+        role = "SPAR_TSC_ADMIN",
+        crudAccess = {'C', 'R', 'U', 'D'}),
+    @AccessLevel(
+        role = "SPAR_MINISTRY_ORCHARD",
+        crudAccess = {'C', 'R', 'U', 'D'}),
+    @AccessLevel(
+        role = "SPAR_NONMINISTRY_ORCHARD",
+        crudAccess = {'C', 'R', 'U', 'D'})
+  })
+  @AccessLevelRequired({'C', 'R'})
```
[the remark this answers is not in the document]
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/FavouriteActivityEndpoint.java:97]: `AccessLevelRequired` describes which operations are made by the endpoint. We need this information to be able to **cross** with roles and its permissions.

File: backend/src/main/java/ca/bc/gov/backendstartapi/config/SecurityConfig.java
Comment (craigyu) [inline backend/src/main/java/ca/bc/gov/backendstartapi/config/SecurityConfig.java:1]: Does the system store the forest client id for a given role? 
A user can have multiple client_ids with each client_id having many roles.
We get these two formats of roles from FAM:
1. `{Role}_{forest_client_id}` (Currently only for `SPAR_NONMINISTRY_ORCHARD`)
2. `{Role}` (Currently only for `SPAR_TSC_ADMIN` and `SPAR_MINISTRY_ORCHARD`)

With the current implementation it seems like option 1's entire string will be considered as a single role

Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/config/SecurityConfig.java:1]: For example, on our front-end we have this way of storing client_id roles

```
[
  {client_id}: [ list of roles ],
  ...
]
```

**Review by craigyu — COMMENTED** [review]
Great changes 💪 I left a question regarding the role processing design

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/FavouriteActivityEndpoint.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/FavouriteActivityEndpoint.java]: After discussing how to simplify this configs, the team decided to drop this complex annotations keeping only one to include the required roles for that endpoint.

File: backend/src/main/java/ca/bc/gov/backendstartapi/config/SecurityConfig.java
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/config/SecurityConfig.java:1]: Great question Craig! No, we're not storing any client id. The interceptor is cutting it off and storing only the role name.  And for the duplicates (same role, different client id), we can trust the `Set` collection of Java, which keeps no duplicates.

but do we need to store the client ids on the user information?

**Review by craigyu — APPROVED** [review]

**Pull request merged**